File: LaRC/CAL_LID_L2_PSCMask_Standard_V2_00_2021_03_26T00_00_00ZN_hdf.m
%  This example code illustrates how to access and visualize LaRC CALIPSO LIDAR
% Level 2 PCSMask HDF4 file in MATLAB. 
%
%  If you have any questions, suggestions, comments on this example, 
% please use the HDF-EOS Forum (http://hdfeos.org/forums). 
% 
%  If you would like to see an  example of any other NASA HDF/HDF-EOS
% data product that is not listed in the HDF-EOS  Comprehensive Examples 
% page (http://hdfeos.org/zoo), feel free to contact us at 
% eoshelp@hdfgroup.org or post it at the HDF-EOS Forum 
% (http://hdfeos.org/forums).
%
% Usage:save this script and run (without .m at the end)
%
%  $matlab -nosplash -nodesktop -r CAL_LID_L2_PSCMask_Standard_V2_00_2021_03_26T00_00_00ZN_hdf
%
% Tested under: MATLAB R2021a
% Last updated: 2022-10-04

import matlab.io.hdf4.*

% Open the HDF4 File.
FILE_NAME = 'CAL_LID_L2_PSCMask-Standard-V2-00.2021-03-26T00-00-00ZN.hdf';
SD_id = sd.start(FILE_NAME, 'rdonly');

% Read data.
datafield_name='PSC_Feature_Mask';
sds_index = sd.nameToIndex(SD_id, datafield_name);
sds_id = sd.select(SD_id, sds_index);
data = sd.readData(sds_id);
sd.endAccess(sds_id);

% Read _FillValue from data field.
fillvalue_index = sd.findAttr(sds_id, 'fillvalue ');
fillvalue = str2double(sd.readAttr(sds_id, fillvalue_index));

% Read units attribute from data field.
units_index = sd.findAttr(sds_id, 'units ');
units = sd.readAttr(sds_id, units_index);

% Read lat.
lat_name='Latitude';
sds_index = sd.nameToIndex(SD_id, lat_name);
sds_id = sd.select(SD_id, sds_index);
lat = sd.readData(sds_id);
sd.endAccess(sds_id);

% Read lon.
lon_name='Longitude';
sds_index = sd.nameToIndex(SD_id, lon_name);
sds_id = sd.select(SD_id, sds_index);
lon = sd.readData(sds_id);
sd.endAccess(sds_id);

% Read altitude.
alt_name='Altitude';
sds_index = sd.nameToIndex(SD_id, alt_name);
sds_id = sd.select(SD_id, sds_index);
alt = sd.readData(sds_id);

units_index = sd.findAttr(sds_id, 'units ');
units_alt = sd.readAttr(sds_id, units_index);

sd.endAccess(sds_id);

% Close the file.
sd.close(SD_id);

% Convert data to double type for plot.
data=double(data);
alt=double(alt);
lat=double(lat);
lon=double(lon);

% Subset data at a certain altitude.
data = data(121, :);

% Replace the fill value with NaN.
data(data==fillvalue) = NaN;

% Create a figure to plot the data.
f = figure('Name', FILE_NAME, ...
    'Renderer', 'zbuffer', ...
    'Position', [0,0,800,600], ...
    'Visible', 'off', ...
    'PaperPositionMode', 'auto');
axesm('MapProjection','eqdcylin', ...
      'Frame','on','Grid','on', ...
      'MeridianLabel','on','ParallelLabel','on');
scatterm(lat, lon, 1, data(:));
colormap('Jet');
h = colorbar();

coast = load('coastlines.mat');
plotm(coast.coastlat,coast.coastlon,'k')

tstring = {FILE_NAME;[datafield_name, ' at Altitude = ', num2str(alt(121)), ...
                    ' (', units_alt, ')']};
title(tstring, 'Interpreter', 'none', 'FontSize', 16, ...
      'FontWeight','bold');
set(get(h, 'title'), 'string', units, 'FontSize',12,'FontWeight','bold');
saveas(f, [FILE_NAME '.m.png']);
exit;

File: LaRC/CAL_LID_L2_PSCMask-Standard-V2-00.2021-03-26T00-00-00ZN.hdf.py
"""
This example code illustrates how to access and visualize a LaRC CALIPSO
LIDAR L2 PSCMask HDF4 file in Python.

If you have any questions, suggestions, or comments on this example, please use
the HDF-EOS Forum (http://hdfeos.org/forums).  If you would like to see an
example of any other NASA HDF/HDF-EOS data product that is not listed in the
HDF-EOS Comprehensive Examples page (http://hdfeos.org/zoo), feel free to
contact us at eoshelp@hdfgroup.org or post it at the HDF-EOS Forum
(http://hdfeos.org/forums).

Usage:  save this script and run

    $python CAL_LID_L2_PSCMask-Standard-V2-00.2021-03-26T00-00-00ZN.hdf.py

The HDF file must be in your current working directory.

Tested under: Python 3.9.12 :: Miniconda
Last updated: 2022-09-29
"""
import os
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt

from pyhdf.SD import SD, SDC
from matplotlib import colors
from mpl_toolkits.basemap import Basemap

FILE_NAME = 'CAL_LID_L2_PSCMask-Standard-V2-00.2021-03-26T00-00-00ZN.hdf'

# Read data field.
DATAFIELD_NAME = 'PSC_Feature_Mask'

hdf = SD(FILE_NAME, SDC.READ)
        
# Read dataset.
data2D = hdf.select(DATAFIELD_NAME)
data = data2D[:,:].astype(np.float64)

# Retrieve the attributes.
attrs = data2D.attributes(full=1)
fva=attrs["fillvalue "]
fill_value = fva[0]
ua=attrs["units "]
units = ua[0]

 # Replace the missing values with NaN.        
data[data == float(fill_value)] = np.nan
datam = np.ma.masked_array(data, np.isnan(data))
    
# Read geolocation datasets.
latitude = hdf.select('Latitude')
lat = latitude[:]

longitude = hdf.select('Longitude')
lon = longitude[:]

altitude = hdf.select('Altitude')
alt = altitude[:]
attrs = altitude.attributes(full=1)
ua=attrs["units "]
alt_units = ua[0]

m = Basemap(projection='cyl', resolution='l',
                llcrnrlat=-90, urcrnrlat=90,
                llcrnrlon=-180, urcrnrlon=180)
m.drawcoastlines(linewidth=0.5)
m.drawparallels(np.arange(-90, 91, 30), labels=[1, 0, 0, 0])
m.drawmeridians(np.arange(-180, 181, 45), labels=[0, 0, 0, 1])
sc = m.scatter(lon, lat, c=datam[:,120], s=1,
               cmap=plt.cm.jet,
               edgecolors=None, linewidth=0)
cb = m.colorbar()
cb.set_label(units)

basename = os.path.basename(FILE_NAME)
long_name = DATAFIELD_NAME + ' at Altitude=' + str(alt[120]) + alt_units
plt.title('{0}\n{1}'.format(basename, long_name))
fig = plt.gcf()
pngfile = "{0}.py.png".format(basename)
fig.savefig(pngfile)
    

 
